File: 0130-surrounded-regions/0130-surrounded-regions.java
class Solution {

    int[][] directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};

    class Barrier {
        int[] parent;
        int[] rank;

        public Barrier(char[][] board) {
            int m = board.length;
            int n = board[0].length;
            parent = new int[m * n];
            rank = new int[m * n];
            for (int i = 0; i < m; i++) {
                for (int j = 0; j < n; j++) {
                    if (board[i][j] == 'O') {
                        parent[i * n + j] = i * n + j;
                        rank[i * n + j] = 1;
                    }
                }
            }
        }

        public int find(int x) {
            if (parent[x] != x) {
                parent[x] = find(parent[x]);
            }
            return parent[x];
        }

        public void union(int x, int y) {
            int rootX = find(x);
            int rootY = find(y);
            if (rank[rootX] < rank[rootY]) {
                parent[rootX] = rootY;
            } else if (rank[rootX] > rank[rootY]) {
                parent[rootY] = rootX;
            } else {
                parent[rootY] = rootX;
                rank[rootX]++;
            }
        }

    }

    public void solve(char[][] board) {
        Barrier barrier = new Barrier(board);
        HashSet<Integer> unBoards = new HashSet<>();
        // 记录包含边界的并查集，其他全部刷成X
        int m = board.length;
        int n = board[0].length;
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (board[i][j] == 'O') {
                    for (int k = 0; k < 4; k++) {
                        int x = i + directions[k][0];
                        int y = j + directions[k][1];
                        if (x >= 0 && x < m && y >= 0 && y < n && board[x][y] == 'O') {
                            barrier.union(i * n + j, x * n + y);
                        }
                    }
                    if (i == 0 || i == m - 1 || j == 0 || j == n - 1) {
                        unBoards.add(barrier.find(i * n + j));
                    }
                }
            }
        }
        // 开始刷被包围的点
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (board[i][j] == 'O' && !unBoards.contains(barrier.find(i * n + j))) {
                    board[i][j] = 'X';
                }
            }
        }
    }
}

File: 0130-surrounded-regions/0130-surrounded-regions.py
class DSU:
    def __init__(self,n : int):
        self.pa = list(range(n))
        self.rn = [1]*n

    def find(self,x):
        if self.pa[x] != x:
            self.pa[x] = self.find(self.pa[x])
        return self.pa[x]

    def union(self,x,y):
        x,y = self.find(x),self.find(y)
        if x == y :
            return
        if self.rn[x] < self.rn[y]:
            x, y = y , x
        self.pa[y] = x
        self.rn[x] += self.rn[y]

     




class Solution:
    def solve(self, board: List[List[str]]) -> None:
        candidate = []
        m = len(board)
        n = len(board[0])

        def node(x,y):
            return x*n + y

        dsu = DSU(m*n + 1)
        boundary = m*n
        dirs = [(1,0), (-1,0), (0,1), (0,-1)]

        #check through the matrix and construct the array
        #attach the boundary points to a special set

        for i in range(m):
            for j in range(n):
                if board[i][j] == 'O':
                    if i == 0 or i == m - 1 or j == 0 or j == n - 1:
                        dsu.union(node(i,j),boundary)
                    else:
                        for x , y in dirs:
                            ni , nj = i + x , j + y
                            if board[ni][nj] == 'O':
                                dsu.union(node(ni,nj), node(i,j))
        blank = dsu.find(boundary)

        for i in range(m):
            for j in range(n):
                if board[i][j] == 'O':
                    if dsu.find(node(i,j)) != blank:
                        board[i][j] = 'X'

                        

        
        




        